packages/flutter_tools/lib/src/flx.dart

```diff
@@ -75,26 +75,79 @@ Future<Null> build({
 
   DevFSContent kernelContent;
   if (!precompiledSnapshot && previewDart2) {
+    final File fingerprintFile = fs.file('$depfilePath.fingerprint');
+    final List<String> inputPaths = <String>[
+      mainPath,
+    ];
+    final List<String> outputPaths = <String>[
+      applicationKernelFilePath,
+    ];
+
+    bool needBuild = true;
+    final List<File> fingerprintFiles = <File>[fingerprintFile, fs.file(depfilePath)]
+      ..addAll(inputPaths.map(fs.file))
+      ..addAll(outputPaths.map(fs.file));
+
+    Future<Fingerprint> makeFingerprint() async {
+      final Set<String> compilerInputPaths = await readDepfile(depfilePath)
+        ..add(mainPath)
+        ..addAll(outputPaths);
+      final Map<String, String> properties = <String, String>{
+        'entryPoint': mainPath,
+      };
+      return new Fingerprint.fromBuildInputs(properties, compilerInputPaths);
+    }
+
+    if (fingerprintFiles.every((File file) => file.existsSync())) {
+      try {
+        final String json = await fingerprintFile.readAsString();
+        final Fingerprint oldFingerprint = new Fingerprint.fromJson(json);
+        if (oldFingerprint == await makeFingerprint()) {
+          needBuild = false;
+          printStatus('Skipping compilation. Fingerprint match.');
+        }
+      } catch (e) {
+        printTrace('Rebuilding kernel file due to fingerprint check error: $e');
+      }
+    }
+
     ensureDirectoryExists(applicationKernelFilePath);
+    final String kernelBinaryFilename =
+      needBuild ?
+          await compile(
+            sdkRoot: artifacts.getArtifactPath(Artifact.flutterPatchedSdkPath),
+            incrementalCompilerByteStorePath: fs.path.absolute(getIncrementalCompilerByteStoreDirectory()),
+            mainPath: fs.file(mainPath).absolute.path,
+            outputFilePath: applicationKernelFilePath,
+            depFilePath: depfilePath,
+            trackWidgetCreation: trackWidgetCreation,
+            fileSystemRoots: fileSystemRoots,
+            fileSystemScheme: fileSystemScheme,
+            packagesPath: packagesPath,
+          )
+          : applicationKernelFilePath;
+
 
-    final String kernelBinaryFilename = await compile(
-      sdkRoot: artifacts.getArtifactPath(Artifact.flutterPatchedSdkPath),
-      incrementalCompilerByteStorePath: fs.path.absolute(getIncrementalCompilerByteStoreDirectory()),
-      mainPath: fs.file(mainPath).absolute.path,
-      outputFilePath: applicationKernelFilePath,
-      depFilePath: depfilePath,
-      trackWidgetCreation: trackWidgetCreation,
-      fileSystemRoots: fileSystemRoots,
-      fileSystemScheme: fileSystemScheme,
-      packagesPath: packagesPath,
-    );
     if (kernelBinaryFilename == null) {
-      throwToolExit('Compiler terminated unexpectedly on $mainPath');
+      throwToolExit('Compiler failed on $mainPath');
```

Comment:
Isn't this check of 'kernelBinaryFilename == null' needed only when the compilation is done, in the non compilation path it could just be an assertion correct because we expect to find the file?

Reply:
Refactored.